cmd/terramate/cli/script_info.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2023 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package cli
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/terramate-io/terramate/config"
+	"github.com/terramate-io/terramate/hcl"
+	"github.com/terramate-io/terramate/hcl/ast"
+	"golang.org/x/exp/slices"
+)
+
+func (c *cli) printScriptInfo() {
+	labels := c.parsedArgs.Experimental.Script.Info.Labels
+
+	stacks, err := c.computeSelectedStacks(false)
+	if err != nil {
+		fatal(err, "computing selected stacks")
+	}
+
+	m := newScriptsMatcher(labels)
+	m.Search(c.cfg().Tree(), stacks)
+
+	for _, x := range m.Results {
+		c.output.MsgStdOut("Definition: %v", x.ScriptCfg.Range)
+		c.output.MsgStdOut("Description: %v", formatScriptDescription(x.ScriptCfg))
+		c.output.MsgStdOut("Stacks:")
+		for _, st := range x.Stacks {
+			c.output.MsgStdOut("  %v", st.Dir())
+		}
+		c.output.MsgStdOut("Jobs:")
+		for _, job := range x.ScriptCfg.Jobs {
+			for cmdIdx, cmd := range formatScriptJob(job) {
+				if cmdIdx == 0 {
+					c.output.MsgStdOut("  * %v", cmd)
+				} else {
+					c.output.MsgStdOut("    %v", cmd)
+				}
+			}
+		}
+
+		c.output.MsgStdOut("")
+	}
+}
+
+type scriptInfoEntry struct {
+	ScriptCfg *hcl.Script
+	Stacks    config.List[*config.SortableStack]
+}
+
+type scriptsMatcher struct {
+	labels []string
+
+	Results []*scriptInfoEntry
+}
+
+func newScriptsMatcher(labels []string) *scriptsMatcher {
+	return &scriptsMatcher{
+		labels:  labels,
+		Results: nil,
+	}
+}
+
+func (m *scriptsMatcher) Search(cfg *config.Tree, stacks config.List[*config.SortableStack]) {
+	rootStacks := make(config.List[*config.SortableStack], len(stacks))
+	copy(rootStacks, stacks)
+
+	m.searchRecursive(cfg, &rootStacks)
+}
+
+func tryGetScriptByName(cfg *config.Tree, labels []string) *hcl.Script {
+	for _, s := range cfg.Node.Scripts {
+		if slices.Equal(s.Labels, labels) {
+			return s
+		}
+	}
+	return nil
+}
+
+func sortedKeys[V any](m map[string]V) []string {
+	r := make([]string, 0, len(m))
+	for k := range m {
+		r = append(r, k)
+	}
+	sort.Strings(r)
+	return r
+}
+
+func (m *scriptsMatcher) searchRecursive(cfg *config.Tree, parentStacks *config.List[*config.SortableStack]) {
+	scriptCfg := tryGetScriptByName(cfg, m.labels)
+	if scriptCfg != nil {
+		m.addResult(cfg, scriptCfg, parentStacks)
+	} else {
+		for _, k := range sortedKeys(cfg.Children) {
+			m.searchRecursive(cfg.Children[k], parentStacks)
+		}
+	}
+}
+
+func (m *scriptsMatcher) addResult(cfg *config.Tree, scriptCfg *hcl.Script, parentStacks *config.List[*config.SortableStack]) {
+	newParentStacks := config.List[*config.SortableStack]{}
+	myStacks := config.List[*config.SortableStack]{}
+
+	cfgdir := cfg.Dir()
+	for _, st := range *parentStacks {
+		stackdir := st.Dir()
+
+		if stackdir.HasDirPrefix(cfgdir.String()) {
+			myStacks = append(myStacks, st)
+		} else {
+			newParentStacks = append(newParentStacks, st)
+		}
+	}
```

Comment:
I think the names `parentStacks`, `newParentStacks` and `myStacks` are misleading here.
We start with "all stacks" (in the first call, but later it's the "pending stacks", and for the given `cfg` node, we split it into "child stacks" (or stacks in current script scope) and "unrelated stacks".

The "unrelated stacks" are the stacks that the current script (defined at `cfg`) does not apply, then you update the "pending stacks" pointer to become the rest.

Then down there you call `m.searchRecursive()` again for each cfg children node, providing a pointer to the "child stacks" of this node.

Then, IMO, the names should be:

```
parentStacks    = pendingStacks
newParentStacks = unrelatedStacks
myStacks        = scopedStacks
```

wdyt?

Reply:
I changed it to use the term scope (rootScope, innerScope, outerScope), which should be a better description of the what it actually is.